src/Workspaces/Core/Portable/Storage/AbstractPersistentStorageService.cs

```diff
@@ -44,55 +43,57 @@ protected AbstractPersistentStorageService(IPersistentStorageConfiguration confi
     public ValueTask<IChecksummedPersistentStorage> GetStorageAsync(SolutionKey solutionKey, CancellationToken cancellationToken)
     {
         return solutionKey.FilePath == null
-            ? new(NoOpPersistentStorage.GetOrThrow(Configuration.ThrowOnFailure))
+            ? new(NoOpPersistentStorage.GetOrThrow(solutionKey, Configuration.ThrowOnFailure))
             : GetStorageWorkerAsync(solutionKey, cancellationToken);
     }
 
     internal async ValueTask<IChecksummedPersistentStorage> GetStorageWorkerAsync(SolutionKey solutionKey, CancellationToken cancellationToken)
     {
-        using (await _lock.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+        // Without taking the lock, see if we can use the last storage system we were asked to create.  Ensure we use a
+        // using so that if we don't take it we still release this reference count.
+        using var current = _currentPersistentStorage?.TryAddReference();
+        if (current != null && solutionKey == current.Target.SolutionKey)
         {
-            // Do we already have storage for this?
-            if (solutionKey.Id == _currentPersistentStorageSolutionId)
-            {
-                // We do, great. Increment our ref count for our caller.  They'll decrement it
-                // when done with it.
-                return PersistentStorageReferenceCountedDisposableWrapper.AddReferenceCountToAndCreateWrapper(_currentPersistentStorage!);
-            }
+            // Success, we can use the current storage system.  Ensure we increment the reference count again, so that
+            // this stays alive for the caller when the above reference count drops.
+            return PersistentStorageReferenceCountedDisposableWrapper.AddReferenceCountToAndCreateWrapper(current);
+        }
 
-            var workingFolder = Configuration.TryGetStorageLocation(solutionKey);
-            if (workingFolder == null)
-                return NoOpPersistentStorage.GetOrThrow(Configuration.ThrowOnFailure);
+        var workingFolder = Configuration.TryGetStorageLocation(solutionKey);
+        if (workingFolder == null)
+            return NoOpPersistentStorage.GetOrThrow(solutionKey, Configuration.ThrowOnFailure);
 
-            // If we already had some previous cached service, let's let it start cleaning up
-            if (_currentPersistentStorage != null)
+        ReferenceCountedDisposable<IChecksummedPersistentStorage>? storageToDispose = null;
+        IChecksummedPersistentStorage persistentStorage;
+        using (await _lock.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+        {
+            if (solutionKey != _currentPersistentStorage?.Target?.SolutionKey)
```

Comment:
>? [](http://example.com/codeflow?start=64&length=1)

'?' needed?

Reply:
yes.  _currentPersistentStorage could be null.  if not null, it could point at null.

Reply:
ignore me.